### ISSUE TITLE
Customize metrics: binary_accuracy of outputs > 0.5

### ISSUE THREAD

**Issue by zeka0** [fchollet/keras#6838]
I have searched high and low but found no satisfying results...
So, here is the problem:
The neural network outputs a (4,) array, and the predicted labels should be those greater than 0.5.
The problem is, top_k_categorical_accuracy won't work.(of course)
And, when I tried to implement one by my own,
I noticed that:
`def customized_metrics(y_true, y_pred):`
the parameter **y_true and y_pred** are tensors. And what I tried to do is in-place assignment, making those values greater than 0.5 become 1 and others 0.
Any ideas?

**Comment by td2014** [comment]
Have you checked into the lambda layers?:
For example:
# add a x -> x^2 layer
model.add(Lambda(lambda x: x ** 2))

Documentation link here for details:
https://keras.io/layers/core/

Hope that helps.

**Comment by zeka0** [comment]
I solved it, here is the code to turn output into 1s(if >= 0.5) and 0s(if < 0.5):
```python
def threshold_binarize(x, threshold=0.5):
    ge = tf.greater_equal(x, tf.constant(threshold))
    y = tf.where(ge, x=tf.ones_like(x), y=tf.zeros_like(x))
    return y
```
Lamba layer isn't elegant, and I encountered several mysterious errors using them....

**Comment by zeka0** [comment]
I came across another problem...
While training the model using Adam optimizer and with loss function being 'binary_crossentropy',
I noticed the loss is increasing while metrics(accuracy) increasing too.
What the hell? I thought optimizer should minimize the entropy of the model...

**Comment by Hhhhhhhhhhao** [comment]
@zeka0  how do you solve it finally?

**Comment by hbb21st** [comment]
If so, the threshold_binary_accuracy  doesn't decrease anymore? So how to modify the loss function accordingly?  Still use "binary_crossentropy" like this 
model.compile(loss='binary_crossentropy', optimizer=adam, metrics=[ threshold_binary_accuracy, precision, recall, f1_score ] )
output are:
 300s - loss: 1.3017 - threshold_binary_accuracy: 0.3165 - val_loss: 1.3797 - val_threshold_binary_accuracy: 0.3421

Epoch 00001: threshold_binary_accuracy improved from -inf to 0.31647, saving model to (200,200).h5
Epoch 2/20
 - 291s - loss: 1.2290 - threshold_binary_accuracy: 0.3033 - val_loss: 1.3601 - val_threshold_binary_accuracy: 0.3564

Epoch 00002: threshold_binary_accuracy did not improve from 0.31647
Epoch 3/20
 - 291s - loss: 1.1486 - threshold_binary_accuracy: 0.2963 - val_loss: 1.2142 - val_threshold_binary_accuracy: 0.3058

Epoch 00003: threshold_binary_accuracy did not improve from 0.31647
Epoch 4/20
 - 289s - loss: 1.0801 - threshold_binary_accuracy: 0.2767 - val_loss: 1.2386 - val_threshold_binary_accuracy: 0.3089

Epoch 00004: threshold_binary_accuracy did not improve from 0.31647
Epoch 5/20
 - 289s - loss: 1.0360 - threshold_binary_accuracy: 0.2696 - val_loss: 1.2030 - val_threshold_binary_accuracy: 0.3042

Epoch 00005: threshold_binary_accuracy did not improve from 0.31647
Epoch 6/20
 - 290s - loss: 0.9817 - threshold_binary_accuracy: 0.2606 - val_loss: 1.1841 - val_threshold_binary_accuracy: 0.3249

Epoch 00006: threshold_binary_accuracy did not improve from 0.31647
Epoch 7/20
 - 287s - loss: 0.9713 - threshold_binary_accuracy: 0.2573 - val_loss: 1.1359 - val_threshold_binary_accuracy: 0.3060

Epoch 00007: threshold_binary_accuracy did not improve from 0.31647
Epoch 8/20
 - 286s - loss: 0.9159 - threshold_binary_accuracy: 0.2410 - val_loss: 1.2064 - val_threshold_binary_accuracy: 0.2783
